File: evals/retrievers/bm25.ts
import type { Chunk } from '../lib/corpus.ts'
import { tokenize } from '../lib/tokenize.ts'
import type { Retriever } from './keyword.ts'

const K1 = 1.2
const B = 0.75

/**
 * The tokenizer is a parameter so a stemmed variant is the same retriever over different terms —
 * anything else would compare two implementations rather than the one thing under test.
 */
export function bm25Retriever(
  chunks: Chunk[],
  opts: { name?: string; tokenizer?: (text: string) => string[] } = {},
): Retriever {
  const { name = 'bm25', tokenizer = tokenize } = opts
  const docs = chunks.map((c) => ({ id: c.id, terms: tokenizer(c.text) }))
  const avgLen = docs.reduce((s, d) => s + d.terms.length, 0) / (docs.length || 1)

  const df = new Map<string, number>()
  const tf: Array<Map<string, number>> = docs.map((d) => {
    const counts = new Map<string, number>()
    for (const t of d.terms) {
      counts.set(t, (counts.get(t) ?? 0) + 1)
    }
    for (const t of counts.keys()) {
      df.set(t, (df.get(t) ?? 0) + 1)
    }
    return counts
  })

  const N = docs.length
  return {
    name,
    search(query, k) {
      const terms = tokenizer(query)
      const scored = docs.map((d, i) => {
        let score = 0
        for (const t of terms) {
          const f = tf[i].get(t)
          if (!f) {
            continue
          }
          const n = df.get(t) ?? 0
          const idf = Math.log(1 + (N - n + 0.5) / (n + 0.5))
          const norm = f + K1 * (1 - B + (B * d.terms.length) / avgLen)
          score += idf * ((f * (K1 + 1)) / norm)
        }
        return { id: d.id, score }
      })
      return scored
        .filter((s) => s.score > 0)
        .sort((a, b) => b.score - a.score)
        .slice(0, k)
        .map((s) => s.id)
    },
  }
}

File: evals/retrievers/bm25.py
import math
from collections import Counter
from collections.abc import Callable

from lib.corpus import Chunk
from lib.tokenize import tokenize

Tokenizer = Callable[[str], list[str]]

K1 = 1.2
B = 0.75


class BM25Retriever:
    def __init__(self, chunks: list[Chunk], name: str = 'bm25', tokenizer: Tokenizer = tokenize) -> None:
        self.name = name
        self.tokenizer = tokenizer
        self.docs = [(c.id, tokenizer(c.text)) for c in chunks]
        self.avg_len = sum(len(terms) for _, terms in self.docs) / (len(self.docs) or 1)
        self.tf = [Counter(terms) for _, terms in self.docs]
        self.df = Counter(term for counts in self.tf for term in counts)
        self.n = len(self.docs)

    def search(self, query: str, k: int) -> list[str]:
        terms = self.tokenizer(query)
        scored: list[tuple[str, float]] = []
        for i, (id, doc_terms) in enumerate(self.docs):
            score = 0.0
            for term in terms:
                f = self.tf[i][term]
                if not f:
                    continue
                n = self.df[term]
                idf = math.log(1 + (self.n - n + 0.5) / (n + 0.5))
                norm = f + K1 * (1 - B + (B * len(doc_terms)) / self.avg_len)
                score += idf * ((f * (K1 + 1)) / norm)
            scored.append((id, score))
        hits = [(id, score) for id, score in scored if score > 0]
        ranked = sorted(hits, key=lambda pair: pair[1], reverse=True)
        return [id for id, _ in ranked[:k]]
